docs/design-documents/20231117-better-processors.md

```diff
@@ -0,0 +1,889 @@
+without being processed.
+
+This way we can simplify the `filter` processor and, more importantly, give the user
+the control of defining when a processor should be executed.
+
+We propose adding a field `condition` in the processor configuration that allows you to
+write a Go template that returns a boolean. If the returned value evaluates to `true`,
+the processor will be executed, otherwise it will be skipped.
+If the condition field is empty it is the same as setting the condition to `true`.
+In other words, by default all records are passed to the processor.
+
+```yaml
+processors:
+  id:   example
+  type: custom.javascript
+  # Only execute processor if the metadata field "opencdc.version" contains value "v1"
+  condition: {{ eq .Metadata.opencdc.version "v1" }}
+  settings:
+    script: >
+      function process(record) {
+        logger.Info().Msg("executing JavaScript processor on OpenCDC v1 record");
+        return record;
+      }
+```
+
+### Addressing record fields
+
+Anytime a processor contains a setting that allows you to pick a field or write a
+condition we recommend using the same syntax as the Go stdlib package
+[text/template](https://pkg.go.dev/text/template). This way we are leaning on syntax
+that is thoroughly documented and has a lot of resources online outside of the
+Conduit ecosystem. Additionally, we already use the same approach in specifying the
+record [output format](https://conduit.io/docs/connectors/output-formats).
+
+Here are a few examples of addressing record fields:
+
+- `.` - references the whole record
+- `.Payload.After` - references the field that normally carries the payload
+- `{{ eq .Key.foo "123" }}` - evaluates to `true` if the key contains the value
+  "123" in field `foo`
+
+Note that referencing a field does not require the expression to be nested inside of
+brackets (`{{...}}`), as we don't want the expression to be evaluated. We are simply
+referencing a field with the same syntax we would in a Go template. On the other hand,
+when we **do** want the expression to be evaluated, we need to put the expression in
+brackets. This allows us to differentiate between static values and expressions.
+
+Go templates evaluate to a string, therefore we need to think about how we
+plan to handle cases when we don't expect a string as an output. For instance,
+conditions need to evaluate to a boolean value. We propose to simply use the package
+`strconv`, specifically the function
+[`strconv.ParseBool`](https://pkg.go.dev/strconv#ParseBool) to convert the result to
+a boolean value. This means that values 1, t, T, TRUE, true, True are converted to
+`true`. In addition, any value that is not 0, f, F, FALSE, false, False will return
+an error, which we should also regard as `false`, but additionally log a warning.
+
+Note that we still need to evaluate if `text/template` is actually the implementation
+we want to use to evaluate these expressions. If the performance is too slow we might
```

Comment:
I do have a feeling that the template matching may add some latency, though it does depend on how well the template machine performs.